cmd/ipfs/main.go

```diff
@@ -473,6 +481,9 @@ func (i *cmdInvocation) setupInterruptHandler() {
 	sig := allInterruptSignals()
 
 	go func() {
+		// wait till the context is ready to be closed
+		<-ctx.ContextIsReadyToBeClosed
```

Comment:
By the way, we can't `close(ctx.initDone)` before [this line](https://github.com/kkoroviev/go-ipfs/blob/fix-daemon-races/cmd/ipfs/daemon.go#L272) because the context is read [here](https://github.com/kkoroviev/go-ipfs/blob/fix-daemon-races/cmd/ipfs/daemon.go#L265) and can be modified in `setupInterruptHandler()`.